GetIntoTeachingApiTests/Services/CrmCacheTests.cs

```diff
@@ -0,0 +1,65 @@
+﻿using System;
+using FluentAssertions;
+using GetIntoTeachingApi.Services;
+using GetIntoTeachingApiTests.Utils;
+using Microsoft.Extensions.Logging;
+using Moq;
+using Xunit;
+
+namespace GetIntoTeachingApiTests.Services
+{
+    public class CrmCacheTests
+    {
+        private readonly Mock<ILogger<CrmCache>> _mockLogger;
+        private readonly CrmCache _cache;
+
+        public CrmCacheTests()
+        {
+            _mockLogger = new Mock<ILogger<CrmCache>>();
+            _cache = new CrmCache(_mockLogger.Object);
+        }
+
+        [Fact]
+        public void GetOrCreate_WhenStoreIsEmpty_Creates()
+        {
+            var result = _cache.GetOrCreate("key", DateTime.Now.AddSeconds(30), () => "value");
+
+            result.Should().Be("value");
+        }
+
+        [Fact]
+        public void GetOrCreate_WhenEntryExpired_Refreshes()
+        {
+            _cache.GetOrCreate("key", DateTime.Now.AddSeconds(-30), () => "value");
+
+            var result = _cache.GetOrCreate("key", DateTime.Now.AddSeconds(30), () => "new-value");
+
+            result.Should().Be("new-value");
+        }
+
+        [Fact]
+        public void GetOrCreate_WhenEntryFresh_Gets()
+        {
+            _cache.GetOrCreate("key", DateTime.Now.AddSeconds(30), () => "value");
+
+            var result = _cache.GetOrCreate("key", DateTime.Now.AddSeconds(30), () => "new-value");
+
+            result.Should().Be("value");
+        }
+
+        [Fact]
+        public void GetOrCreate_WhenRefreshFails_ReturnsStaleData()
+        {
+            _cache.GetOrCreate("key", DateTime.Now.AddSeconds(-30), () => "value");
+
+            var result = _cache.GetOrCreate("key", DateTime.Now.AddSeconds(30), () =>
+            { 
+                throw new Exception("bang");
+                return "";

```

Comment:
this return is triggering some warnings

Reply:
Yeah this can be ignored, it's in the test suite and I need to test the exception being raised in a functional argument, don't think there's a way to do it without the warning but I'll take another look and pick it up in my next PR if there is